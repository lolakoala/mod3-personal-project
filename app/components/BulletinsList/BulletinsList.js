import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

class BulletinsList extends Component {
  getINeedToRead = (bulletins, id) => {
    return bulletins.filter(bulletin => {
      return !bulletin.hasRead.includes(id);
    }).sort((itemA, itemB) => Date.parse(itemA.datePosted) - Date.parse(itemB.datePosted));
  }

  getWeekBulletins = bulletins => {
    return bulletins.filter(bulletin => Date.now() - bulletin.id <= 604800000)
      .sort((itemA, itemB) => Date.parse(itemA.datePosted) - Date.parse(itemB.datePosted));
  }

  getMatchingBulletins = (bulletins, value) => {
    return bulletins.filter(bulletin => {
      return bulletin.title.includes(value) || bulletin.details.includes(value);
    }).sort((itemA, itemB) => Date.parse(itemB.datePosted) - Date.parse(itemA.datePosted));
  }

  getBulletins = () => {
    const { usersHouse, currentUser, placeRendered, searchValue } = this.props;
    if (placeRendered === '/') {
      return this.getINeedToRead(usersHouse.bulletins, currentUser.id);
    } else if (placeRendered === '/houselist') {
      return usersHouse.bulletins.sort((itemA, itemB) => Date.parse(itemB.datePosted) - Date.parse(itemA.datePosted));
    } else if (placeRendered === 'summary') {
      return this.getWeekBulletins(usersHouse.bulletins);
    } else if (placeRendered === 'search') {
      return this.getMatchingBulletins(usersHouse.bulletins, searchValue);
    }
  }

  getError = () => {
    const { placeRendered, usersHouse } = this.props;
    if (placeRendered === '/') {
      return 'You have no unread bulletins.';
    } else if (placeRendered === '/houselist') {
      return `${usersHouse.houseName} has no bulletins posted.`;
    } else if (placeRendered === 'summary') {
      return `${usersHouse.houseName} has no bulletins posted this week.`;
    } else if (placeRendered === 'search') {
      return 'No bulletins match your query.';
    }
  }

  render() {
    const { usersHouse, currentUser, addReaderToBulletin, placeRendered } = this.props;
    if (usersHouse.bulletins.length && usersHouse.bulletins[0].title !== 'fake') {
      const bulletins = this.getBulletins();
      if (bulletins.length) {
        return (
          <div className='bulletinslist'>
            {placeRendered === 'summary' ? <h4>Current Bulletins</h4> : null}
            {placeRendered === '/' ? <h4>Bulletins I Need to Read</h4> : null}
            {bulletins.map(bulletin => {
              let bulletinClass;
              bulletin.hasRead.includes(currentUser.id) ? bulletinClass = 'read' : bulletinClass = 'not-read';
              return (<div key={bulletin.id} className={`${bulletinClass} bulletin`}>
                <Link
                  className='bulletin-title'
                  to={`bulletins/${bulletin.id}`}
                  onClick={() => addReaderToBulletin(bulletin.id, currentUser.id, usersHouse)}>
                  {bulletin.title}
                </Link>
                <p>{`Posted on: ${bulletin.datePosted}`}</p>
              </div>);
            })}
          </div>
        );
      } else {
        return <p className='error'>{this.getError()}</p>;
      }
    } else {
      return <p className='error'>{`${usersHouse.houseName} has no bulletins posted.`}</p>;
    }

  }
}

export default BulletinsList;

BulletinsList.propTypes = {
  usersHouse: PropTypes.object,
  currentUser: PropTypes.object,
  addReaderToBulletin: PropTypes.func,
  placeRendered: PropTypes.string,
  searchValue: PropTypes.string
};
