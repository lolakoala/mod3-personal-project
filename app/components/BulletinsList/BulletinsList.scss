.bulletinslist h4,
.error {
  background-color: rgba(141, 132, 73, 0.7);
  color: #3d6aa2;
  font-family: 'Shadows Into Light', cursive;
  font-size: 20px;
  width: 250px;
  padding: 5px;
  text-align: center;
  height: 25px;
  border-radius: 5px;
  line-height: 35px;
  margin-top: 20px;
}

.bulletinslist .bulletin {
  display: flex;
  height: 30px;
  width: 300px;
  border-radius: 5px;
  border: 2px solid rgba(141, 132, 73, 0.7);
  margin-top: 10px;
  line-height: 30px;
  // text-align: center;
  font-family: 'BenchNine', sans-serif;
  position: relative;
  margin-top: 10px;
}

.bulletin a {
  border: 2px solid #D83D59;
  text-decoration: none;
  width: 100px;
  overflow: hidden;
  display: inline-block;
  margin: 2px 5px;
  background-color: #D83D59;
  height: 25px;
  color: #F4CCE2;
  font-size: 20px;
  border-radius: 25px;
  text-align: center;
}

.bulletin a:hover {
  cursor: pointer;
  border: 2px dashed #D83D59;
  background-color: #F4CCE2;
  color: #D83D59;
}

.bulletin p {
  width: 180px;
  font-size: 15px;
  display: inline-block;
  margin: 2px 0px 2px 5px;
  text-align: center;
}
