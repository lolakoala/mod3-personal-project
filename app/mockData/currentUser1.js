const currentUser1 = {
  id : "RVde1m9BhkQBLZlTJNsFlQZXqP43",
  name: "Amanda Brenner",
  house: "Dog9"
};

export default currentUser1;
