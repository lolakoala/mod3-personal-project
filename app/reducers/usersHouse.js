import firebase from '../firebase.js';

const usersHouse = (state = {}, action) => {
  const housesRef = firebase.database().ref('houses');
  switch (action.type) {
  case 'CREATE_HOUSE':
    housesRef.push(action.house);
    return action.house;
  case 'GET_HOUSE':
    const updatedHouse = Object.assign(action.usersHouse, { users: [...action.usersHouse.users, action.user] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updatedHouse);
    return updatedHouse;
  case 'LOGIN_SUCCESS':
    return action.usersHouse;
  case 'SIGNOUT':
    return {};
  case 'LEAVE_HOUSE':
    const deleteUser = action.usersHouse.users.filter(user => user.id !== action.currentUser.id);
    const houseWithoutUser = Object.assign(action.usersHouse, { users: deleteUser });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithoutUser);
    return {};
  case 'ADD_BILL':
    const takeOutFakeBills = action.usersHouse.bills.filter(bill => bill.title !== 'fake');
    const houseWithBill = Object.assign(action.usersHouse, { bills: [...takeOutFakeBills, action.bill] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithBill);
    return houseWithBill;
  case 'ADD_BULLETIN':
    const takeOutFakeBulletins = action.usersHouse.bulletins.filter(bulletin => bulletin.title !== 'fake');
    const houseWithBulletin = Object.assign(action.usersHouse, { bulletins: [...takeOutFakeBulletins, action.bulletin] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithBulletin);
    return houseWithBulletin;
  case 'ADD_CHORE':
    const takeOutFakeChore = action.usersHouse.chores.filter(chore => chore.title !== 'fake');
    const houseWithChore = Object.assign(action.usersHouse, { chores: [...takeOutFakeChore, action.chore] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithChore);
    return houseWithChore;
  case 'ADD_READER':
    const oldBulletin = action.usersHouse.bulletins.find(bulletin => bulletin.id === action.bulletinId);
    if (oldBulletin.hasRead.includes(action.userId)) {
      return action.usersHouse;
    }
    const newBulletin = Object.assign(oldBulletin, { hasRead: [...oldBulletin.hasRead, action.userId] });
    const updateBulletins = action.usersHouse.bulletins.filter(bulletin => {
      return bulletin.id !== action.bulletinId;
    });
    const updateWithBulletin = Object.assign(action.usersHouse, { bulletins: [...updateBulletins, newBulletin] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updateWithBulletin);
    return updateWithBulletin;
  case 'MARK_PAID':
    const oldBill = action.usersHouse.bills.find(bill => bill.id === action.billId);
    const user = oldBill.allUsersTotals.find(user => user.id === action.userId);
    if (user.paid === true) {
      return action.usersHouse;
    }
    const newUser = Object.assign(user, { paid: true });
    const newAllUsersTotals = oldBill.allUsersTotals.filter(user => user.id !== action.userId);
    const newBill = Object.assign(oldBill, { allUsersTotals: [...newAllUsersTotals, newUser] });
    const updatedBills = action.usersHouse.bills.filter(bill => {
      return bill.id !== action.billId;
    });
    const updateWithBill = Object.assign(action.usersHouse, { bills: [...updatedBills, newBill] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updateWithBill);
    return updateWithBill;
  default:
    return state;
  }
};

export default usersHouse;
