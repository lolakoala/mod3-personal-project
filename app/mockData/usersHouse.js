import currentUser1 from './currentUser1.js';
import currentUser2 from './currentUser2.js';

const bill1 = {
  allUsersTotals: [
    {
      id: currentUser1.id,
      paid: false,
      total: 50
    },
    {
      id: currentUser2.id,
      paid: false,
      total: 50
    }
  ],
  datePosted: "11/01/2017",
  details: "We used a lot of water.",
  duedate: "11/9/17",
  id: "b3019762705779",
  parsedDuedate: 1510210800000,
  postedBy: currentUser1,
  title: "Water Bill",
  total: "100"
};

const bulletin1 = {
  datePosted: "11/01/2017",
  details: "The dogs got out through a hole in the fence yesterday.",
  hasRead: [currentUser1.id],
  id: 1509567098747,
  postedBy: currentUser1,
  title: "Don't leave dogs out unattended."
};

const chore1 = {
  assignedTo: currentUser1.id,
  datePosted: "10/31/2017",
  details: "Empty the dishwasher.",
  done: false,
  id: "c1509485241076",
  postedBy: currentUser1,
  title: "Dishes",
  urgency: "low"
};

const usersHouse = {
  houseKey: "-KxoIBWy5UYRscFYdWty",
  bills: [bill1],
  bulletins: [bulletin1],
  chores: [chore1],
  houseCode: "woof9",
  houseName: "Dog9",
  users: [currentUser1, currentUser2]
};

export default usersHouse;
