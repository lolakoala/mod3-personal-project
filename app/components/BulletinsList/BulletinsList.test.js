import React from "react";
import BulletinsList from './BulletinsList.js';
import { shallow } from "enzyme";
import usersHouse from '../../mockData/usersHouse.js';
import currentUser1 from '../../mockData/currentUser1.js';

describe('BulletinsList component', () => {
  const mockFn = jest.fn();
  const wrapper = shallow(<BulletinsList usersHouse={usersHouse}
    currentUser={currentUser1}
    addReaderToBulletin={mockFn}
    placeRendered={'/houselist'}
    searchValue={''}/>);

  it('should exist', () => {
    expect(wrapper).toBeDefined();
  });

  it('should have headers', () => {
    const headers = wrapper.find('.bulletin-headers');

    expect(headers.length).toEqual(1);
  });

  it('should have a title column header', () => {
    const titleHeader = wrapper.find('.title-header');

    expect(titleHeader.text()).toEqual('Title');
  });

  it('should have a dateposted column header', () => {
    const datepostedHeader = wrapper.find('.date-header');

    expect(datepostedHeader.text()).toEqual('Date Posted');
  });

  it('should have a all read header', () => {
    const allReadHeader = wrapper.find('.allread-header');

    expect(allReadHeader.text()).toEqual('All Read');
  });

  it('should have display all bulletins of the house', () => {
    const bulletins = wrapper.find('.bulletin');

    expect(bulletins.length).toEqual(usersHouse.bulletins.length);
  });

  it('should fire action', () => {
    const addReader = wrapper.find('.bulletin-title').first();

    addReader.simulate('click');
    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
