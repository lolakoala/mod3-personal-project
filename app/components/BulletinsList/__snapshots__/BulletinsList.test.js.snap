// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulletinsList component should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BulletinsList
    addReaderToBulletin={[Function]}
    currentUser={
      Object {
        "house": "Dog9",
        "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
        "name": "Amanda Brenner",
      }
    }
    placeRendered="/houselist"
    searchValue=""
    usersHouse={
      Object {
        "bills": Array [
          Object {
            "allUsersTotals": Array [
              Object {
                "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
                "paid": false,
                "total": 50,
              },
              Object {
                "id": "cFNdorl5ceRyMr9XGvPFRVEcw7I3",
                "paid": false,
                "total": 50,
              },
            ],
            "datePosted": "11/01/2017",
            "details": "We used a lot of water.",
            "duedate": "11/9/17",
            "id": "b3019762705779",
            "parsedDuedate": 1510210800000,
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Water Bill",
            "total": "100",
          },
        ],
        "bulletins": Array [
          Object {
            "datePosted": "11/01/2017",
            "details": "The dogs got out through a hole in the fence yesterday.",
            "hasRead": Array [
              "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            ],
            "id": 1509567098747,
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Don't leave dogs out unattended.",
          },
        ],
        "chores": Array [
          Object {
            "assignedTo": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            "datePosted": "10/31/2017",
            "details": "Empty the dishwasher.",
            "done": false,
            "id": "c1509485241076",
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Dishes",
            "urgency": "low",
          },
        ],
        "houseCode": "woof9",
        "houseKey": "-KxoIBWy5UYRscFYdWty",
        "houseName": "Dog9",
        "users": Array [
          Object {
            "house": "Dog9",
            "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            "name": "Amanda Brenner",
          },
          Object {
            "house": "Dog9",
            "id": "cFNdorl5ceRyMr9XGvPFRVEcw7I3",
            "name": "Lola Koala",
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        null,
        null,
        Array [
          <div
            className="read bulletin"
          >
            <Link
              className="bulletin-title"
              onClick={[Function]}
              replace={false}
              to="bulletins/1509567098747"
            >
              Don't leave dogs out unattended.
            </Link>
            <p>
              Posted on: 11/01/2017
            </p>
          </div>,
        ],
      ],
      "className": "bulletinslist",
    },
    "ref": null,
    "rendered": Array [
      null,
      null,
      Object {
        "instance": null,
        "key": "1509567098747",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Link
              className="bulletin-title"
              onClick={[Function]}
              replace={false}
              to="bulletins/1509567098747"
            >
              Don't leave dogs out unattended.
            </Link>,
            <p>
              Posted on: 11/01/2017
            </p>,
          ],
          "className": "read bulletin",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "class",
            "props": Object {
              "children": "Don't leave dogs out unattended.",
              "className": "bulletin-title",
              "onClick": [Function],
              "replace": false,
              "to": "bulletins/1509567098747",
            },
            "ref": null,
            "rendered": "Don't leave dogs out unattended.",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Posted on: 11/01/2017",
            },
            "ref": null,
            "rendered": "Posted on: 11/01/2017",
            "type": "p",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          null,
          null,
          Array [
            <div
              className="read bulletin"
            >
              <Link
                className="bulletin-title"
                onClick={[Function]}
                replace={false}
                to="bulletins/1509567098747"
              >
                Don't leave dogs out unattended.
              </Link>
              <p>
                Posted on: 11/01/2017
              </p>
            </div>,
          ],
        ],
        "className": "bulletinslist",
      },
      "ref": null,
      "rendered": Array [
        null,
        null,
        Object {
          "instance": null,
          "key": "1509567098747",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                className="bulletin-title"
                onClick={[Function]}
                replace={false}
                to="bulletins/1509567098747"
              >
                Don't leave dogs out unattended.
              </Link>,
              <p>
                Posted on: 11/01/2017
              </p>,
            ],
            "className": "read bulletin",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "class",
              "props": Object {
                "children": "Don't leave dogs out unattended.",
                "className": "bulletin-title",
                "onClick": [Function],
                "replace": false,
                "to": "bulletins/1509567098747",
              },
              "ref": null,
              "rendered": "Don't leave dogs out unattended.",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Posted on: 11/01/2017",
              },
              "ref": null,
              "rendered": "Posted on: 11/01/2017",
              "type": "p",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
